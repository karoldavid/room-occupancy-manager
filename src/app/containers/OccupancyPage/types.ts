export enum RoomType {
  PREMIUM = 'premium',
  ECONOMY = 'economy',
}

export enum GuestsErrorType {
  RESPONSE_ERROR = 1,
  GUESTS_NOT_FOUND = 2,
}

export interface AvailableRooms {
  [name: string]: number;
}

export interface AvailableRoomsFormValues {
  [name: string]: {
    value: number;
  };
}

export interface Revenue {
  free: number;
  usage: number;
  total: number;
  currency: string;
}

export type RevenueRoomType = {
  [key in RoomType]: Revenue;
};

export interface BookingState {
  availableRooms: AvailableRooms;
  guests: number[];
}

export interface OccupancyState extends BookingState {
  loading: boolean;
  error?: GuestsErrorType | null;
}

export type ContainerState = OccupancyState;
