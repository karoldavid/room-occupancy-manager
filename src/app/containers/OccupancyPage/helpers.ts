import sortBy from 'lodash/sortBy';
import {
  AvailableRooms,
  RoomType,
  RevenueRoomType,
  BookingState,
} from './types';

const MAX_AMOUNT_EURO_PREMIUM = 100;

export const initialRevenueRoomState: RevenueRoomType = {
  [RoomType.PREMIUM]: { usage: 0, total: 0, currency: 'EURO' },
  [RoomType.ECONOMY]: { usage: 0, total: 0, currency: 'EURO' },
};

export const revenueRoomReducer = (
  occupancyState: BookingState,
  revenueRoomState: RevenueRoomType,
): RevenueRoomType => {
  return sortBy(occupancyState.guests)
    .reverse()
    .reduce(
      (prev: RevenueRoomType, curr: number, index: number): RevenueRoomType => {
        const newOccupation = { ...prev };

        if (
          (curr >= MAX_AMOUNT_EURO_PREMIUM &&
            occupancyState.availableRooms.premium >
              newOccupation.premium.usage) ||
          (curr < MAX_AMOUNT_EURO_PREMIUM &&
            occupancyState.availableRooms.premium >
              newOccupation.premium.usage &&
            occupancyState.availableRooms.economy <
              occupancyState.guests.length - index + 1)
        ) {
          newOccupation.premium.usage += 1;
          newOccupation.premium.total += curr;
        } else if (
          curr < MAX_AMOUNT_EURO_PREMIUM &&
          occupancyState.availableRooms.economy > newOccupation.economy.usage
        ) {
          newOccupation.economy.usage += 1;
          newOccupation.economy.total += curr;
        }

        return {
          ...newOccupation,
        };
      },
      revenueRoomState,
    );
};

export const getDetailsDataByCategory = (category: RoomType) => (
  revenueRoom: RevenueRoomType,
  availableRooms: AvailableRooms,
) => ({
  ...revenueRoom[category],
  free: availableRooms[category],
});
