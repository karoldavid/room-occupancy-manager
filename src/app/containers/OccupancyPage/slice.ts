import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import {
  ContainerState,
  AvailableRooms,
  RoomType,
  GuestsErrorType,
} from './types';

export const initialState: ContainerState = {
  availableRooms: { [RoomType.PREMIUM]: 0, [RoomType.ECONOMY]: 0 },
  guests: [],
  loading: false,
  error: null,
};

const occupancySlice = createSlice({
  name: 'occupancy',
  initialState,
  reducers: {
    updateAvailableRoomsForm(state, action: PayloadAction<AvailableRooms>) {
      state.availableRooms = action.payload;
    },
    loadGuests(state) {
      state.guests = [];
      state.loading = true;
      state.error = null;
    },
    guestsLoaded(state, action: PayloadAction<number[]>) {
      state.guests = action.payload;
      state.loading = false;
      state.error = null;
    },
    guestsError(state, action: PayloadAction<GuestsErrorType>) {
      state.error = action.payload;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = occupancySlice;
