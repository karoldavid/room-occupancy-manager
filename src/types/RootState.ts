import { OccupancyState } from 'app/containers/OccupancyPage/types';

export interface RootState {
  occupancy: OccupancyState;
}
