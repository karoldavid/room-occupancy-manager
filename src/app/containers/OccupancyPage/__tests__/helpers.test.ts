import { revenueRoomReducer, initialRevenueRoomState } from '../helpers';
import { BookingState, RevenueRoomType, RoomType } from '../types';
import { initialState } from '../slice';
import data from '../mock/guests.json';

describe('revenueRoomReducer', () => {
  let occupancyState: BookingState;
  let initialRevenueRoom: RevenueRoomType;

  beforeEach(() => {
    occupancyState = {
      ...initialState,
      guests: data,
    };
    initialRevenueRoom = {
      [RoomType.PREMIUM]: { usage: 0, total: 0, currency: 'EURO' },
      [RoomType.ECONOMY]: { usage: 0, total: 0, currency: 'EURO' },
    };
  });

  it('should calculate the correct revenue and room occupancy 1', () => {
    const free = { premium: 3, economy: 3 };

    const newOccupancy = {
      ...occupancyState,
      availableRooms: {
        premium: free.premium,
        economy: free.economy,
      },
    };

    const expectedResult = {
      premium: {
        usage: 3,
        total: 738,
        currency: 'EURO',
      },
      economy: {
        usage: 3,
        total: 167,
        currency: 'EURO',
      },
    };

    occupancyState.availableRooms.premium = free.premium;
    occupancyState.availableRooms.economy = free.economy;

    const result = revenueRoomReducer(newOccupancy, initialRevenueRoom);

    expect(result).toEqual(expectedResult);
  });

  it('should calculate the correct revenue and room occupancy 2', () => {
    const free = { premium: 7, economy: 5 };

    const newOccupancy = {
      ...occupancyState,
      availableRooms: {
        premium: free.premium,
        economy: free.economy,
      },
    };

    const expectedResult = {
      premium: {
        usage: 6,
        total: 1054,
        currency: 'EURO',
      },
      economy: {
        usage: 4,
        total: 189,
        currency: 'EURO',
      },
    };

    const result = revenueRoomReducer(newOccupancy, {
      ...initialRevenueRoom,
    });
    expect(result).toEqual(expectedResult);
  });

  it('should calculate the correct revenue and room occupancy 3', () => {
    const free = { premium: 2, economy: 7 };

    const newOccupancy = {
      ...occupancyState,
      availableRooms: {
        premium: free.premium,
        economy: free.economy,
      },
    };

    const expectedResult = {
      premium: {
        usage: 2,
        total: 583,
        currency: 'EURO',
      },
      economy: {
        usage: 4,
        total: 189,
        currency: 'EURO',
      },
    };

    const result = revenueRoomReducer(newOccupancy, {
      ...initialRevenueRoomState,
    });

    expect(result).toEqual(expectedResult);
  });

  it('should calculate the correct revenue and room occupancy 4', () => {
    const free = { premium: 7, economy: 1 };

    const newOccupancy = {
      ...occupancyState,
      availableRooms: {
        premium: free.premium,
        economy: free.economy,
      },
    };

    const expectedResult = {
      premium: {
        usage: 7,
        total: 1153,
        currency: 'EURO',
      },
      economy: {
        usage: 1,
        total: 45,
        currency: 'EURO',
      },
    };

    const result = revenueRoomReducer(newOccupancy, {
      ...initialRevenueRoom,
    });

    expect(result).toEqual(expectedResult);
  });
});
