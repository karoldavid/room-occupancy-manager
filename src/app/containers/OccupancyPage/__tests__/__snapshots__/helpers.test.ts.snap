// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers getDetailsDataByCategory should match snapshots 1`] = `
Object {
  "currency": "EURO",
  "free": 3,
  "total": 738,
  "usage": 3,
}
`;

exports[`helpers getDetailsDataByCategory should match snapshots 2`] = `
Object {
  "currency": "EURO",
  "free": 3,
  "total": 167,
  "usage": 3,
}
`;
